File: backend/app/services/session_service.py
from uuid import UUID

from backend.app.models.session import Session, SessionCreate, SessionStatus
from backend.app.repositories.session_repository import SessionRepository


class SessionService:
    """
    Business logic for Sessions.
    """

    def __init__(self, repository: SessionRepository):
        self.repository = repository

    def create_session(self, data: SessionCreate) -> Session:
        session = Session(
            title=data.title,
            summary=data.summary,
            location=data.location,
        )
        return self.repository.create(session)

    def list_sessions(self) -> list[Session]:
        return self.repository.list_all()

    def get_session(self, session_id: UUID) -> Session | None:
        return self.repository.get_by_id(session_id)

    def complete_session(self, session_id: UUID) -> Session | None:
        session = self.repository.get_by_id(session_id)

        if session is None:
            return None

        session.status = SessionStatus.COMPLETED
        return self.repository.update(session)

File: frontend/lib/services/session_service.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/session.dart';
import 'api.dart';

class SessionService {
  Future<List<NexusSession>> getSessions() async {
    final response = await http.get(
      Uri.parse('${Api.baseUrl}/sessions'),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to load sessions: ${response.statusCode}',
      );
    }

    final List<dynamic> data = jsonDecode(response.body);

    return data
        .map(
          (item) => NexusSession.fromJson(
            item as Map<String, dynamic>,
          ),
        )
        .toList();
  }

  Future<NexusSession> createSession({
    required String title,
    String? summary,
    String? location,
    String? workspaceId,
  }) async {
    final response = await http.post(
      Uri.parse('${Api.baseUrl}/sessions'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'title': title,
        'summary': summary,
        'location': location,
        'workspace_id': workspaceId,
      }),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to create session: ${response.statusCode}',
      );
    }

    return NexusSession.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }

  Future<NexusSession> completeSession(
    String sessionId,
  ) async {
    final response = await http.post(
      Uri.parse(
        '${Api.baseUrl}/sessions/$sessionId/complete',
      ),
    );

    if (response.statusCode != 200) {
      throw Exception(
        'Failed to complete session: ${response.statusCode}',
      );
    }

    return NexusSession.fromJson(
      jsonDecode(response.body) as Map<String, dynamic>,
    );
  }
}